src/puffin/src/puffin_manager/cache_manager/moka_cache_manager.rs

```diff
@@ -0,0 +1,941 @@
+                let path = self.base_dir.join(&file_name);
+
+                let size = Self::write_blob(&path, &init_fn).await?;
+
+                let guard = Arc::new(FsBlobGuard {
+                    path,
+                    delete_queue: self.delete_queue.clone(),
+                });
+                Ok(CacheValue::File { guard, size })
+            })
+            .await
+            .context(CacheGetSnafu)?;
+
+        match v {
+            CacheValue::File { guard, .. } => Ok(guard),
+            _ => unreachable!(),
+        }
+    }
+
+    async fn get_dir<'a>(
+        &self,
+        puffin_file_name: &str,
+        key: &str,
+        init_fn: Box<dyn InitDirFn + Send + Sync + 'a>,
+    ) -> Result<Self::Dir> {
+        let cache_key = Self::encode_cache_key(puffin_file_name, key);
+
+        let v = self
+            .cache
+            .try_get_with(cache_key.clone(), async {
+                if let Some(v) = self.recycle_bin.remove(&cache_key).await {
+                    return Ok(v);
+                }
+
+                let dir_name = format!("{}.{}", cache_key, uuid::Uuid::new_v4());
+                let path = self.base_dir.join(&dir_name);
+
+                let size = Self::write_dir(&path, &init_fn).await?;
+
+                let guard = Arc::new(FsDirGuard {
+                    path,
+                    delete_queue: self.delete_queue.clone(),
+                });
+                Ok(CacheValue::Dir { guard, size })
+            })
+            .await
+            .context(CacheGetSnafu)?;
+
+        match v {
+            CacheValue::Dir { guard, .. } => Ok(guard),
+            _ => unreachable!(),
+        }
+    }
+
+    async fn put_dir(
+        &self,
+        puffin_file_name: &str,
+        key: &str,
+        dir_path: PathBuf,
+        size: u64,
+    ) -> Result<()> {
+        let cache_key = Self::encode_cache_key(puffin_file_name, key);
+
+        self.cache
+            .try_get_with(cache_key.clone(), async move {
+                if let Some(v) = self.recycle_bin.remove(&cache_key).await {
+                    return Ok(v);
+                }
+
+                let dir_name = format!("{}.{}", cache_key, uuid::Uuid::new_v4());
+                let path = self.base_dir.join(&dir_name);
+
+                fs::rename(&dir_path, &path).await.context(RenameSnafu)?;
+
+                let guard = Arc::new(FsDirGuard {
+                    path,
+                    delete_queue: self.delete_queue.clone(),
+                });
+                Ok(CacheValue::Dir { guard, size })
+            })
+            .await
+            .map(|_| ())
+            .context(CacheGetSnafu)
+    }
+}
+
+impl MokaCacheManager {
+    fn encode_cache_key(puffin_file_name: &str, key: &str) -> String {
+        let mut hasher = Sha256::new();
+        hasher.update(puffin_file_name);
+        hasher.update(key);
+        hasher.update(puffin_file_name);
```

Comment:
Why do we update the puffin file name twice?

Reply:
It can avoid conflicts caused by different `(puffin_file_name, key)` but the same after concatenation.

Reply:
Could you please document this in the comment?